### PR TITLE
adding randomized algorithms

### PR DESCRIPTION
**Pull request by deneick**
I added randomized range finder algorithms: one adaptive algorithm and one with power iterations.

### CODE REVIEW

**Comment by pmli** [comment]
Hi @deneick, this is looking good.

Is there some literature you can reference? If yes, it would go in `docs/source/bibliography.rst`, which you can then cite in the function docstring.

**Comment by deneick** [comment]
Hi @pmli, I added the literature.

File: src/pymor/algorithms/randrangefinder.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    
+    if source_product is None:
+        lambda_min = 1
+    elif issparse(source_product.matrix):
+        lambda_min = eigsh(source_product.matrix, which="SM", return_eigenvectors=False, k=1)[0]
+    else:
+        lambda_min = eigh(source_product.matrix.todense(), eigvals_only=True)[0]
+    
+    testlimit = erfinv( (failure_tolerance / min(A.source.dim, A.range.dim))**(1./num_testvecs)) * tol * np.sqrt(2. * lambda_min)
+    maxnorm = float('inf')
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: I guess `maxnorm = np.inf` is fine.

File: src/pymor/algorithms/randrangefinder.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        M.append(A.apply(r))
+    
+    if source_product is None:
+        lambda_min = 1
+    elif issparse(source_product.matrix):
+        lambda_min = eigsh(source_product.matrix, which="SM", return_eigenvectors=False, k=1)[0]
+    else:
+        lambda_min = eigh(source_product.matrix.todense(), eigvals_only=True)[0]
+    
+    testlimit = erfinv( (failure_tolerance / min(A.source.dim, A.range.dim))**(1./num_testvecs)) * tol * np.sqrt(2. * lambda_min)
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: This line is a bit longish... Breaking it into at least two would be nice.

File: src/pymor/algorithms/randrangefinder.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    maxnorm = float('inf')
+    
+    while(maxnorm > testlimit):
+        basis_length = len(B)
+        v = A.source.random(distribution='normal')
+        if iscomplex:
+            v += 1j*A.source.random(distribution='normal')
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: Similar to above, you do not have to generate column by column.

File: src/pymor/algorithms/randrangefinder.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        B.append(A.apply(v))
+        gram_schmidt(B, range_product, offset=basis_length, copy=False)
+        M -= B.lincomb(B.inner(M,range_product).T)
+        maxnorm = np.max(np.abs(M.norm(range_product)))
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: I guess `np.abs` is not necessary.

File: src/pymor/algorithms/randrangefinder.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    testlimit = erfinv( (failure_tolerance / min(A.source.dim, A.range.dim))**(1./num_testvecs)) * tol * np.sqrt(2. * lambda_min)
+    maxnorm = float('inf')
+    
+    while(maxnorm > testlimit):
+        basis_length = len(B)
+        v = A.source.random(distribution='normal')
+        if iscomplex:
+            v += 1j*A.source.random(distribution='normal')
+        B.append(A.apply(v))
+        gram_schmidt(B, range_product, offset=basis_length, copy=False)
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: I guess here in `gram_schimdt`, `atol` and `rtol` should be set to zero, or they should be optional parameters of `adaptive_rrf`...

File: src/pymor/algorithms/randrangefinder.py (hunk 3 above)
Comment (deneick) [reply src/pymor/algorithms/randrangefinder.py]: Here, in every iteration only one random vector is applied to the operator and appended to the basis.

File: src/pymor/algorithms/randrangefinder.py (hunk 4 above)
Comment (deneick) [reply src/pymor/algorithms/randrangefinder.py]: The return value of `M.norm(range_product)` is complex (about 1e-17) if M is complex. Should I rather use `.real` then?

File: src/pymor/algorithms/randrangefinder.py (hunk 3 above)
Comment (pmli) [reply src/pymor/algorithms/randrangefinder.py]: Ah, right

File: src/pymor/algorithms/randrangefinder.py (hunk 4 above)
Comment (pmli) [reply src/pymor/algorithms/randrangefinder.py]: Well, I guess then `norm` should be fixed...

Comment (pmli) [reply src/pymor/algorithms/randrangefinder.py]: Could you open an issue with a minimal example?

Comment (deneick) [reply src/pymor/algorithms/randrangefinder.py]: Yes, I did

**Comment by deneick** [comment]
I guess I covered everything.

File: src/pymor/algorithms/randrangefinder.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    R = A.source.random(num_testvecs, distribution='normal')
+    if iscomplex:
+        R += 1j*A.source.random(num_testvecs, distribution='normal')
+
+    if source_product is None:
+        lambda_min = 1
+    elif issparse(source_product.matrix):
+        lambda_min = eigsh(source_product.matrix, which="SM", return_eigenvectors=False, k=1)[0]
+    else:
+        lambda_min = eigh(source_product.matrix.todense(), eigvals_only=True)[0]
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: Hm, it seems to me you are assuming here that `source_product` is a `NumpyMatrixOperator`. Is this assumption general enough? If not, you can use `algorithms.to_matrix.to_matrix` to convert the operator into a NumPy/SciPy matrix or, better yet, pass [SciPy's LinearOperator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html#scipy.sparse.linalg.LinearOperator) to `eigsh`.

**Comment by pmli** [comment]
Increasing the test coverage would be nice. You could add some simple tests to `src/pymortests`. E.g.
```python
def test_rrf():
    A = ...
    Q = rrf(A)
    assert Q in A.range
    assert len(Q) == 2
```

**Comment by pmli** [comment]
Just a few minor remarks:
- there is no reference in `approximate_operatornorm`
- there are a few `flake8` issues

Otherwise, looks good to me. Sphinx documentation is rendering nicely and code coverage is excellent (98%).
@pymor/pymor-devs Any other opinions?

**Comment by deneick** [comment]
I don't have a reference for approximate_operatornorm, so I added some explanation.

**Comment by sdrave** [comment]
Sorry, I need some time to think about this, and I don't have any right now. Will try to submit a review by the end of next week.

**Comment by renefritze** [comment]
FYI: Slycot switched their build system to `skbuild` in their latest point release and there seems to be no longer wheels for that release. Hence the install fails since `skbuild` is not a dependency, hence our CI fails. We should just pin it to version 0.3.3. exactly for now.

File: src/pymor/algorithms/randrangefinder.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                 failure_tolerance=1e-15, num_testvecs=20, lambda_min=None, iscomplex=False):
+    r"""Adaptive randomized range approximation of `A`.
+
+    This is an implementation of Algorithm 1 in [BS18]_.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `B` with the property
+
+    .. math::
+        \Vert A - P_{span(B)} A \Vert \leq tol
```
Comment (sdrave) [inline src/pymor/algorithms/randrangefinder.py]: Maybe you should specify the norm here.

File: src/pymor/algorithms/randrangefinder.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    if source_product is None:
+        lambda_min = 1
+    elif lambda_min is None:
+        def mv(v):
+            return source_product.apply(source_product.source.from_numpy(v)).to_numpy()
+
+        def mvinv(v):
+            return source_product.apply_inverse(source_product.range.from_numpy(v)).to_numpy()
+        L = LinearOperator((source_product.source.dim, source_product.range.dim), matvec=mv, matmat=mv)
```
Comment (sdrave) [inline src/pymor/algorithms/randrangefinder.py]: Is `matmat=mv` really correct? I would expect scipy to use column vectors, but `from_numpy`/`to_numpy` uses column vectors (@pmli, I know, I know ...).

File: src/pymor/algorithms/randrangefinder.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        Linv = LinearOperator((source_product.range.dim, source_product.source.dim), matvec=mvinv, matmat=mvinv)
```
Comment (sdrave) [inline src/pymor/algorithms/randrangefinder.py]: Same here.

File: src/pymor/algorithms/randrangefinder.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        lambda_min = eigsh(L, sigma=0, which="LM", return_eigenvectors=False, k=1, OPinv=Linv)[0]
```
Comment (sdrave) [inline src/pymor/algorithms/randrangefinder.py]: I am curious: Have you compared the performance for `sigma=None` vs. `sigma=0`? Probably depends on the problem, though.

File: src/pymor/algorithms/randrangefinder.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+
+@defaults('q', 'l')
+def rrf(A, source_product=None, range_product=None, q=2, l=8, iscomplex=False):
+    """Randomized range approximation of `A`.
+
+    This is an implementation of Algorithm 4.4 in [HMT11]_.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `Q` whose vectors form an orthonomal basis for the range of `A`.
```
Comment (sdrave) [inline src/pymor/algorithms/randrangefinder.py]: `approximate orthonormal basis`

File: src/pymor/algorithms/randrangefinder.py (hunk 8 above)
Comment (deneick) [reply src/pymor/algorithms/randrangefinder.py]: I just realized scipy does not pass `matmat` to arpack in eigsh, but only in svds, so I will leave this out here.

File: src/pymor/algorithms/randrangefinder.py (hunk 10 above)
Comment (deneick) [reply src/pymor/algorithms/randrangefinder.py]: Yes, `sigma=None` is faster for small problems (e.g. `source_product` in `pymortests.randrangefinder`) and `sigma=0` is faster for bigger problems (e.g. `range_product` in `pymortests.randrangefinder`).

File: src/pymor/algorithms/randrangefinder.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        R += 1j*A.source.random(num_testvecs, distribution='normal')
+
+    if source_product is None:
+        lambda_min = 1
+    elif lambda_min is None:
+        def mv(v):
+            return source_product.apply(source_product.source.from_numpy(v)).to_numpy()
+
+        def mvinv(v):
+            return source_product.apply_inverse(source_product.range.from_numpy(v)).to_numpy()
```
Comment (pmli) [inline src/pymor/algorithms/randrangefinder.py]: Is `adaptive_rrf` actually supposed to work if `source_product.space.from_numpy` is not implemented? Or is `from_numpy` always implemented for `source_products` in applications?

Comment (sdrave) [reply src/pymor/algorithms/randrangefinder.py]: > Or is from_numpy always implemented for source_products in applications?

No, this cannot be expected. In that case you would need to specify `lambda_min`. For the future, I would hope that at some point we reimplement something like ARPACK in pyMOR. Not sure how much effort would be required, however.

File: src/pymor/algorithms/randrangefinder.py (hunk 10 above)
Comment (sdrave) [reply src/pymor/algorithms/randrangefinder.py]: Sounds reasonable.

**Pull request merged**